rs/consensus/src/consensus/validator.rs

```diff
@@ -990,6 +990,26 @@ impl Validator {
                 continue;
             }
 
+            // If there exists an equivocation proof for the signer of the
+            // proposal at the same height, we ignore the block. The only
```

Comment:
I wonder if this is safe to do, without making some other changes first.

Suppose I'm a malicious rank-0 block maker and do the following:
1) Propose a valid block and send it *ONLY* to rank-1, rank-2, ..., rank-(f+1) block makers 
2) Create a fake equivocation proof for myself (meaning I set `EquivocationProof::signer` to my node id, and `EquivocationProof::{hash_1, hash_2, signature_1, signature_2}` to some _valid_ rubbish) and send it to all my peers, except the rank-1, rank-2, ..., rank-(f+1) block makers

Now:
1) if `rank-i` block maker validates my block quickly enough, then it won't propose a block themselves (https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/consensus/src/consensus/block_maker.rs?L127). This means that if this happens for all rank block makers, then I'm going to be the only block maker who proposed a block in this round;
2) If a peer validates my equivocation proof before validating my block, then it will not validate my block (unless it's notarized later). If this happens to at least `f` peers, my block will never be notarized.

Am I missing something or this could lead to a stall? Is there some existing mechanism which would prevent this from happening?

Reply:
Good catch, note that the [consensus protocol paper](https://eprint.iacr.org/2021/632.pdf) also states (Section 5.2):
> A party _P_ will “hold back” from [proposing its own block] if there is
an apparently “better” block in its pool. Here, by a “better” block, we mean a block of
lower rank than that of _P_ that has not been disqualified.

So we would probably need to adapt `is_better_block_proposal_available()` accordingly.

On a more general note, should there be some kind of one-pager for this change, or should we review it against the paper?

Reply:
True, whenever we qualify blocks as "valid" or "better" the equivocations has to be considered.

@eichhorl I think we have right now the problem that the logic of validation is sometimes implicit and it scattered over the implementation. If one does not know all of these pieces, there will be no guarantee that the one-pager is exhaustive wrt. scope to be covered right? So maybe we could start with a one pager just as a list of points from the paper matched with the existing implementation? @dist1ll  wdyt?

And a different but related question to everyone: should we even consider an equivocation proof where the signer and the equivocator are the same as valid? This should never happen for any honest replica, so maybe we should consider this case in the equiv. proof validation? Then the attack described by @kpop-dfinity won't be possible.

Reply:
That sounds good, I'll create a one pager so we get the order/scope of the changes right.

Reply:
> should we even consider an equivocation proof where the signer and the equivocator are the same as valid? 

Afaik the proofs are not signed so we can't check that?

> This should never happen for any honest replica, so maybe we should consider this case in the equiv. proof validation? 

True, but then it's enough that there are two colluding malicious nodes and they produce equivocation proofs for each other